api/tx_args.go

```diff
@@ -0,0 +1,304 @@
+// validateArgs checks the validity of SendTxArgs values.
+// The each tx types has its own validation logic to give detailed errors to users.
+func (args *SendTxArgs) validateArgs() error {
+	if args.TypeInt == nil {
+		return errTxArgNilTxType
+	}
+
+	// TODO - Arguments validation will be implemented for each tx type
+	//switch *args.TypeInt {
+	//case types.TxTypeLegacyTransaction:
+	//	if args.Data != nil && args.Input != nil && !bytes.Equal(*args.Data, *args.Input) {
+	//		return errTxArgInvalidInputData
+	//	}
+	//	if args.To == nil {
+	//		// Contract creation
+	//		var input []byte
+	//		if args.Data != nil {
+	//			input = *args.Data
+	//		} else if args.Input != nil {
+	//			input = *args.Input
+	//		}
+	//		if len(input) == 0 {
+	//			return errTxArgNilContractData
+	//		}
+	//	}
+	//case types.TxTypeValueTransfer:
+	//case types.TxTypeFeeDelegatedValueTransfer:
+	//}
+
+	return nil
+}
+
+// genTxValuesMap generates a value map used used in "NewTransactionWithMap" function.
+// This function assigned all non-nil values regardless of the tx type.
+// Invalid values in the map will be validated in "NewTransactionWithMap" function.
+func (args *SendTxArgs) genTxValuesMap() map[types.TxValueKeyType]interface{} {
+	values := make(map[types.TxValueKeyType]interface{})
+
+	// common tx fields.  They should have values after executing "setDefaults" function.
+	if args.TypeInt == nil || args.Nonce == nil || args.Gas == nil || args.GasPrice == nil {
+		return values
+	}
+	values[types.TxValueKeyFrom] = args.From
+	values[types.TxValueKeyNonce] = uint64(*args.Nonce)
+	values[types.TxValueKeyGasLimit] = uint64(*args.Gas)
+	values[types.TxValueKeyGasPrice] = (*big.Int)(args.GasPrice)
+
+	// optional tx fields
+	if args.To != nil {
+		if args.TypeInt.IsContractDeploy() {
+			values[types.TxValueKeyTo] = args.To
+		} else {
+			values[types.TxValueKeyTo] = *args.To
+		}
+	}
+	if args.FeePayer != nil {
+		values[types.TxValueKeyFeePayer] = *args.FeePayer
+	}
+	if args.FeeRatio != nil {
+		values[types.TxValueKeyFeeRatioOfFeePayer] = *args.FeeRatio
+	}
+	if args.Value != nil {
+		values[types.TxValueKeyAmount] = (*big.Int)(args.Value)
+	}
+	if args.Data != nil {
+		values[types.TxValueKeyData] = ([]byte)(*args.Data)
+	}
+	if args.CodeFormat != nil {
+		values[types.TxValueKeyCodeFormat] = *args.CodeFormat
+	}
+	if args.HumanReadable != nil {
+		values[types.TxValueKeyHumanReadable] = *args.HumanReadable
+	}
+	if args.AccountKey != nil {
+		values[types.TxValueKeyAccountKey] = *args.AccountKey
+	}
+
+	return values
+}
+
+// toTransaction returns an unsigned transaction filled with values in SendTxArgs.
+func (args *SendTxArgs) toTransaction() (*types.Transaction, error) {
+	var input []byte
+
+	// provide detailed error messages to users (optional)
+	if err := args.validateArgs(); err != nil {
+		return nil, err
+	}
+
+	// for TxTypeLegacyTransaction
+	if *args.TypeInt == types.TxTypeLegacyTransaction {
+		if args.Data != nil {
+			input = *args.Data
```

Comment:
In here doesn't need to check Input for checking 'errTxArgInvalidInputData' condition?

Reply:
I was planning to add those checks in `validateArgs` which will be implemented soon.
However, the checks should be placed here since `validateArgs` is optional. 
I will edit the code. Thank you.

Reply:
PTAL